File: packages/core/src/actions/query/transaction.ts
import { z } from "zod";
import { ActionHandler } from "../../types.js";

export const TransactionSchema = z.object({
    action: z.literal("transaction"),
    operations: z.array(z.object({
        sql: z.string().describe("SQL statement to execute"),
        params: z.array(z.unknown()).optional().describe("Query parameters"),
    })).min(1).describe("Array of SQL statements to execute atomically in a single transaction"),
});

/**
 * Batch Transaction Handler.
 * Executes multiple statements atomically without persistent session management.
 */
export const transactionHandler: ActionHandler<typeof TransactionSchema> = {
    schema: TransactionSchema,
    handler: async (params, context) => {
        // We use a dedicated session connection just for this batch, then close it.
        // This ensures atomicity and isolation.
        let sessionId: string | undefined;
        try {
            sessionId = await context.sessionManager.createSession();
            const executor = context.sessionManager.getSessionExecutor(sessionId)!;

            await executor.execute("BEGIN");
            const results = [];
            
            let i = 0;
            for (const op of params.operations) {
                try {
                    const res = await executor.execute(op.sql, op.params);
                    results.push(res);
                    i++;
                } catch (error: any) {
                    await executor.execute("ROLLBACK");
                    throw new Error(`Transaction failed at operation ${i}: ${error.message}`);
                }
            }

            await executor.execute("COMMIT");
            return {
                status: "committed",
                results
            };
        } finally {
            // Always close the temporary session if it was created
            if (sessionId) {
                await context.sessionManager.closeSession(sessionId);
            }
        }
    },
};

File: coldquery/actions/query/transaction.py
from typing import Any, Dict, List, Optional
from fastmcp.exceptions import ToolError

from coldquery.core.context import ActionContext
from coldquery.middleware.session_echo import enrich_response


async def transaction_handler(params: Dict[str, Any], context: ActionContext) -> str:
    """Handles the 'transaction' action to execute a batch of SQL queries atomically."""
    operations: Optional[List[Dict[str, Any]]] = params.get("operations")

    if not operations:
        raise ToolError(
            "The 'operations' parameter is required for the 'transaction' action."
        )

    session_id = await context.session_manager.create_session()
    executor = context.session_manager.get_session_executor(session_id)
    if not executor:
        raise RuntimeError("Failed to get session executor.")

    results = []
    try:
        await executor.execute("BEGIN")
        for i, op in enumerate(operations):
            sql = op.get("sql")
            query_params = op.get("params")
            if not sql:
                raise ToolError(f"Operation {i} is missing 'sql'.")
            try:
                result = await executor.execute(sql, query_params)
                results.append(result)
            except Exception as e:
                await executor.execute("ROLLBACK")
                raise RuntimeError(
                    f"Transaction failed at operation {i}: {e}"
                ) from e
        await executor.execute("COMMIT")
        return enrich_response(
            {
                "status": "committed",
                "results": [r.to_dict() for r in results],
            },
            session_id,
            context.session_manager,
        )
    finally:
        await context.session_manager.close_session(session_id)
